### PR TITLE
EL-676: Move model calculations

### PR DESCRIPTION
**Pull request by patrick-laa**
## What

[Link to story](https://dsdmoj.atlassian.net/browse/EL-676)

Move logic out of models

## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

File: app/services/assessors/property_assessor.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      def call(property, remaining_mortgage_allowance, level_of_representation, submission_date)
+        transaction_allowance = calculate_property_transaction_allowance(property, level_of_representation, submission_date)
+        allowable_outstanding_mortgage = calculate_outstanding_mortgage(property, remaining_mortgage_allowance)
+        net_value = calculate_net_value(property, transaction_allowance, allowable_outstanding_mortgage)
+        net_equity = calculate_net_equity(property, net_value)
+        main_home_equity_disregard = calculate_main_home_disregard(property, submission_date)
+        assessed_equity = calculate_assessed_equity(net_equity, main_home_equity_disregard)
+        result = Result.new(transaction_allowance:, allowable_outstanding_mortgage:, net_value:, net_equity:, main_home_equity_disregard:, assessed_equity:)
+        save!(property, result)
+        result
```
Comment (starswan) [inline app/services/assessors/property_assessor.rb:20]: freezing to make immutable is a useful addition. Shame we can't use Data (Ruby 3.2 addition)

File: app/services/calculators/state_benefits_calculator.rb [hunk 2]
```diff
@@ -10,10 +10,17 @@ def call(state_benefits)
...
+            collection: state_benefit.state_benefit_payments,
+          )
+
+          # TODO: Stop persisting this
+          state_benefit.update!(monthly_value:)
+
+          total += monthly_value unless state_benefit.exclude_from_gross_income
         end
-        total
+        total.round(2)
```
Comment (starswan) [inline app/services/calculators/state_benefits_calculator.rb:23]: I'm unsure about this change - where should this rounding be done (if at all?)

File: app/services/decorators/v5/disposable_income_result_decorator.rb [hunk 3]
```diff
@@ -23,10 +23,10 @@ def basic_attributes
           dependant_allowance: summary.dependant_allowance.to_f,
           gross_housing_costs: summary.gross_housing_costs.to_f,
           housing_benefit: summary.housing_benefit.to_f,
-          net_housing_costs: summary.net_housing_costs.to_f,
+          net_housing_costs: summary.net_housing_costs.to_f.round(2),
```
Comment (starswan) [inline app/services/decorators/v5/disposable_income_result_decorator.rb:26]: unsure about this rounding change

File: app/services/calculators/state_benefits_calculator.rb (hunk 2 above)
Comment (patrick-laa) [reply app/services/calculators/state_benefits_calculator.rb]: Ah, so the issue was some floating point rounding errors, and the root cause was some `.to_f` in the MonthlyIncomeConverter. Refactoring was causing some slightly _different_ floating behaviours, and the rounding was my clumsy attempt to fix them. The proper fix, which I've now done, is to stop doing `.to_f` on integers and replace them with `.to_d`

File: app/services/decorators/v5/disposable_income_result_decorator.rb (hunk 3 above)
Comment (patrick-laa) [reply app/services/decorators/v5/disposable_income_result_decorator.rb]: As above

**Review by starswan — APPROVED** [review]

**Pull request merged**